quickwit-search/src/client.rs

```diff
@@ -19,39 +19,132 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+use std::fmt;
 use std::net::SocketAddr;
+use std::sync::Arc;
 
 use http::Uri;
 use tonic::transport::Channel;
 use tonic::transport::Endpoint;
+use tonic::Request;
 
-use quickwit_proto::search_service_client::SearchServiceClient;
+use crate::error::parse_grpc_error;
+use crate::SearchError;
+use crate::SearchService;
 
-#[derive(Debug, Clone)]
-pub struct WrappedSearchServiceClient {
-    client: SearchServiceClient<Channel>,
+/// Impl is an enumlation that meant to manage Quickwit's search service client types.
+#[derive(Clone)]
+enum Impl {
+    Local(Arc<dyn SearchService>),
+    Grpc(quickwit_proto::search_service_client::SearchServiceClient<Channel>),
```

Comment:
Why don't we always use grpc, even on self?

Reply:
It allows for calling directly a `SearchService` trait implementation, which is useful both for mocking but also can be used to bypass the transport when root node equals leaf node.
Right now the client uses this for `MockSearchService` for testing, but in the future, it will be possible to use the service directly without going through gRPC for leaf search to itself like an Elasticsearch.

Reply:
Can't we mock the target node when using grpc?

Reply:
And mock transport? or you'd open a new server for every mock?